rfd/0098-registered-openssh-nodes.md

```diff
@@ -0,0 +1,106 @@
+
+```yaml
+kind: node
+sub_kind: openssh
+metadata:
+  name: 5da56852-2adb-4540-a37c-80790203f6a9
+spec:
+  addr: 1.2.3.4:22
+  hostname: agentless-node
+version: v2
+```
+
+When manually registering an OpenSSH node, create a registered OpenSSH node resource file and create the resource in the cluster with `tctl create --force <node.yml>`. Then follow the steps outlined below in `Manual certificate rotation` to configure the registered OpenSSH node.
+
+### OpenSSH CA
+
+For security related reasons that will be discussed below, a new CA will be added called OpenSSH CA. The OpenSSH CA will be responsible for generating and signing user certificates that will be used to authenticate with registered OpenSSH nodes. The public key of the OpenSSH CA will be copied to all registered OpenSSH nodes and configured as `TrustedUserCAKeys` in `sshd_config`.
+
+#### Manual certificate rotation
+
+Users not using EC2 discovery mode and hosting registered OpenSSH nodes themselves will have to rotate the OpenSSH CA manually and generate new SSH host certificates. They can do so by:
+
+1. Exporting the OpenSSH CA: `tctl auth export --type=openssh | sed "s/cert-authority\ //"`
+2. Write the exported CA to the configuration folder of the node's SSH daemon, ex. `/etc/ssh/` for OpenSSH on Linux
+3. Replace the `TrustedUserCAKeys` setting in `sshd_config` to point to the newly created CA
+4. Generate a new host certificate: `tctl auth sign --host=<addr,hostname> --format=openssh --out=host-cert`
+5. Replace the existing host certificate with the new one on the node and set `HostKey` and `HostCertificate` in `sshd_config`
+
+### RBAC
+
+When a user sends a request to a Proxy to connect to a node, the Proxy will attempt to find a node resource by either its hostname or IP, whichever the user specified. If the resource exists and has the `openssh` `sub_kind`, an RBAC check will be performed. If the resource does not exist or isn't an `openssh` node, the connection flow will continue as normal.
+
+### Security
+
+Both RBAC checks and session recording for registered OpenSSH nodes require that users connect through a Proxy. If users are able to connect to registered OpenSSH nodes directly, they can bypass both features. Currently the User CA public key is copied to OpenSSH nodes and configured to be trusted by `sshd`. This means OpenSSH nodes will accept certificates that are signed by the User CA. The problem with that is when Teleport users authenticate with a cluster, the Auth server replies with a certificate that is signed by the User CA. Users could potentially use this certificate to directly connect to registered OpenSSH nodes.
+
+Using the new OpenSSH CA to sign certificates used to authenticate with registered OpenSSH nodes solves this problem, as Teleport users do not have access to any certificates signed by the OpenSSH CA.
```

Comment:
What about agent forwarding?  The user may not have access to the certificate but consider the following with agent forwarding:
1. A session with agent forwarding is established using teleport.  This session is recorded
2. While the above session is still active, the attacker then uses the agent forwarding to establish a new connection to the server.  If this new connection is established from another connection path (ie not through Teleport, possibly to the server directly after forwarding the agent again to the attackers local system), this connection will completely escape recording.

I assume agent forwarding is just not compatible with this feature?

Reply:
@capnspacehook The concern here is that if the OpenSSH CA signed cert is forwarded, then anyone with access to the agent forwarded socket will have access to any other `sshd` node that trusts OpenSSH CA. Please add to the RFD:
1. Even with the agent forwarding enabled `ssh -A`, the generated certificate signed with OpenSSH CA won't be forwarded.
2. RBAC `allow agent forwarding` option will be respected, and the Teleport administrator can block session forwarded in the agentless.
3. By default, agent forwarding is disabled, and it's not required to use agentless mode.

@jentfoo Correct me if I'm missing something.

Reply:
Added to RFD.